gtsam/hybrid/HybridBayesNet.cpp

```diff
@@ -217,18 +227,160 @@ GaussianBayesNet HybridBayesNet::choose(
+      // BayesTreeOrphanWrapper!
+      throwRuntimeError("HybridBayesNet::assembleTree", f);
+    }
+  }
+
+  GaussianBayesNetValTree resultTree(result, [](const GaussianBayesNet &gbn) {
+    return std::make_pair(gbn, 0.0);
+  });
+  return resultTree;
+}
+
+/* ************************************************************************* */
+AlgebraicDecisionTree<Key> HybridBayesNet::model_selection() const {
+  /*
+      To perform model selection, we need:
+      q(mu; M, Z) * sqrt((2*pi)^n*det(Sigma))
+
+      If q(mu; M, Z) = exp(-error) & k = 1.0 / sqrt((2*pi)^n*det(Sigma))
+      thus, q * sqrt((2*pi)^n*det(Sigma)) = q/k = exp(log(q/k))
+      = exp(log(q) - log(k)) = exp(-error - log(k))
+      = exp(-(error + log(k))),
+      where error is computed at the corresponding MAP point, gbn.error(mu).
+
+      So we compute (error + log(k)) and exponentiate later
+    */
+
+  GaussianBayesNetValTree bnTree = assembleTree();
+
+  GaussianBayesNetValTree bn_error = bnTree.apply(
+      [this](const Assignment<Key> &assignment,
+             const std::pair<GaussianBayesNet, double> &gbnAndValue) {
+        //  Compute the X* of each assignment
+        VectorValues mu = gbnAndValue.first.optimize();
+
+        // mu is empty if gbn had nullptrs
+        if (mu.size() == 0) {
+          return std::make_pair(gbnAndValue.first,
+                                std::numeric_limits<double>::max());
+        }
+
+        // Compute the error for X* and the assignment
+        double error =
+            this->error(HybridValues(mu, DiscreteValues(assignment)));
+
+        return std::make_pair(gbnAndValue.first, error);
+      });
+
+  auto trees = unzip(bn_error);
+  AlgebraicDecisionTree<Key> errorTree = trees.second;
+
+  // Only compute logNormalizationConstant
+  AlgebraicDecisionTree<Key> log_norm_constants = DecisionTree<Key, double>(
+      bnTree, [](const std::pair<GaussianBayesNet, double> &gbnAndValue) {
+        GaussianBayesNet gbn = gbnAndValue.first;
+        if (gbn.size() == 0) {
+          return 0.0;
+        }
+        return gbn.logNormalizationConstant();
+      });
+
+  // Compute model selection term (with help from ADT methods)
+  AlgebraicDecisionTree<Key> model_selection_term =
+      (errorTree + log_norm_constants) * -1;
+
+  double max_log = model_selection_term.max();
+  AlgebraicDecisionTree<Key> model_selection = DecisionTree<Key, double>(
+      model_selection_term,
+      [&max_log](const double &x) { return std::exp(x - max_log); });
+  model_selection = model_selection.normalize(model_selection.sum());
+
+  return model_selection;
+}
+
 /* ************************************************************************* */
 HybridValues HybridBayesNet::optimize() const {
   // Collect all the discrete factors to compute MPE
-  DiscreteBayesNet discrete_bn;
+  DiscreteFactorGraph discrete_fg;
+
+  // Compute model selection term
+  AlgebraicDecisionTree<Key> model_selection_term = model_selection();
```

Comment:
naming convention on variables. `modelSelectionTerms`? This is a global comment on this PR.

Reply:
Done!